tests/test_pattern_matching.py

```diff
@@ -0,0 +1,73 @@
+import pytest
+
+from returns.io import IO, IOFailure, IOResult, IOSuccess
+from returns.maybe import Maybe, Nothing, Some
+from returns.result import Failure, Result, Success
+
+
+@pytest.mark.parametrize('container', [
+    Success(10),
+    Success(42),
+    Failure(RuntimeError()),
+    Failure(Exception()),
+])
+def test_result_pattern_matching(container: Result[int, Exception]):
+    """Ensures ``Result`` containers work properly with pattern matching."""
+    match container:
+        case Success(10):
+            assert isinstance(container, Success)
+            assert container.unwrap() == 10
+        case Success(value):
+            assert isinstance(container, Success)
+            assert value == 42
+            assert container.unwrap() == value
+        case Failure(RuntimeError()):
+            assert isinstance(container, Failure)
+            assert isinstance(container.failure(), RuntimeError)
+        case Failure(_):
+            assert isinstance(container, Failure)
+            assert isinstance(container.failure(), Exception)
+        case _:
+            pytest.fail('Was not matched')
+
+
+@pytest.mark.parametrize('container', [
+    Some('SOME'),
+    Some('THERE IS SOME VALUE'),
+    Nothing,
+])
+def test_maybe_pattern_matching(container: Maybe[str]):
+    """Ensures ``Maybe`` containers work properly with pattern matching."""
+    match container:
+        case Some('SOME'):
+            assert isinstance(container, Some)
+            assert container.unwrap() == 'SOME'
+        case Some(value):
+            assert isinstance(container, Some)
+            assert value == 'THERE IS SOME VALUE'
+            assert container.unwrap() == value
+        case Maybe.empty:
+            assert container is Nothing
+        case _:
+            pytest.fail('Was not matched')
+
+
+@pytest.mark.parametrize('container', [
+    IOSuccess(42.0),
+    IOSuccess(10.0),
+    IOFailure(50),
+])
+def test_ioresult_pattern_matching(container: IOResult[float, int]):
+    """Ensures ``IOResult`` containers work properly with pattern matching."""
+    match container:
+        case IOSuccess(Success(42.0)):
```

Comment:
Is there a way to match this as `IOSuccess(42)`?

Reply:
Yep, I've wrote about this in a previous comment

Reply:
Oh, I see: https://github.com/dry-python/returns/pull/991#discussion_r661976248

Ok, let's keep it this way for now!

Reply:
Ok